clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -3569,6 +3557,69 @@ protected enum DelegateConsumerRecordResult {
     SKIPPED_MESSAGE
   }
 
+  private void recordNearlineLocalBrokerToReadyToServerLatency(
+      String storeName,
+      int versionNumber,
+      PartitionConsumptionState partitionConsumptionState,
+      KafkaMessageEnvelope kafkaMessageEnvelope,
+      LeaderProducedRecordContext leaderProducedRecordContext) {
+    long afterProcessingRecordTimestampMs = System.currentTimeMillis();
+    if (!isUserSystemStore() && isHybridMode() && partitionConsumptionState.hasLagCaughtUp()
+        && (partitionConsumptionState.getLeaderFollowerState() == LeaderFollowerStateType.LEADER
+            || partitionConsumptionState.getLeaderFollowerState() == LeaderFollowerStateType.STANDBY)) {
+      if (partitionConsumptionState.getLeaderFollowerState() == LeaderFollowerStateType.LEADER
+          && leaderProducedRecordContext == null) {
+        LOGGER.error(
+            "LeaderProducedRecordContext is null when PCS state is LEADER for storeName: {},"
+                + " versionNumber: {}, partition: {}",
+            storeName,
+            versionNumber,
+            partitionConsumptionState.getPartition());
+        return;
+      }
+      if (partitionConsumptionState.getLeaderFollowerState() == LeaderFollowerStateType.STANDBY
+          && kafkaMessageEnvelope.getLeaderMetadataFooter() == null) {
+        LOGGER.error(
+            "LeaderMetadataFooter is null when PCS state is STANDBY for storeName: {},"
+                + " versionNumber: {}, partition: {}",
+            storeName,
+            versionNumber,
+            partitionConsumptionState.getPartition());
+        return;
+      }
+      boolean hasCorrespondingUpstreamMessage = false; // No point in recording E2E latency if there is no upstream
+      if (partitionConsumptionState.getLeaderFollowerState() == LeaderFollowerStateType.LEADER) {
+        hasCorrespondingUpstreamMessage = leaderProducedRecordContext.hasCorrespondingUpstreamMessage();
```

Comment:
"internalProcessConsumerRecord" is called inside drainer thread, while the state transition happens in the ingestion thread. I would recommend using the "leaderProducerRecordContext" as source of truth instead of using the leader/follower state, like this:
"""
if (leaderProducerdRecordContext != null) {
    hasCorrespondingUpstreamMessage = leaderProducedRecordContext.hasCorrespondingUpstreamMessage();
} else if (kafkaMessageEnvelope.getLeaderMetadataFooter() != null) {
    kafkaMessageEnvelope.getLeaderMetadataFooter().getUpstreamOffset() >= 0;
}
"""

Reply:
Are you suggesting that the PCS state type is not a reliable indicator of liveness? What I am trying to avoid is measuring latency where we are going from offline to standby / leader but aren't ready to serve yet. The pcs.hasLagCaughtUp should  exclude that situation but I thought additional check on the state would be good. In other words the check is (partitionConsumptionState.getLeaderFollowerState() == LeaderFollowerStateType.LEADER
          && leaderProducedRecordContext == null) which is stronger than just leaderProducedRecordContext == null . Any disadvantage if we test for both?

Reply:
As discussed. Will not be using leader follower state since that can be outdated here.